### DISCUSSION TITLE
Configure netcdf-4.9.2 without libzip

### DISCUSSION THREAD

**Discussion by vbsteph** [Unidata/netcdf-c#2791] [Q&A] [answered]
Attempting to configure netcdf-4.9.2 with Autotools such that `-lzip` does not get included as an "extra library". libzip is installed on the system in question, but due to some downstream linking issues, we specifically want to build netcdf without libzip (or zlib). Based on the docs, I thought that disabling nczarr and netcdf-4 (and therefore hdf5) would be enough, but no luck. I've been unable to come up with a combination of configure line options that eliminates the dependency; this is the most recent set I tried that still resulted in `-lzip` as an extra library:

--disable-shared 
--disable-nczarr 
--disable-hdf5 
--disable-netcdf-4 
--disable-netcdf4 
--disable-dap 
--disable-byterange 
--disable-testsets 
--disable-cdf5

Is there a set of options I can use to eliminate that dependency? Does the configure step always include that library when found regardless of whether it's needed?

EDIT: building with gcc 10.3.0

**Comment by WardF** [comment, marked as answer]
This appears to be an oversight on our end; it is very rare that `libzip` won't be needed to be linked against, but in situations such as yours, that should be possible.  Thank you for pointing out this gap in our build system.  

At current, you may be able to work around this by editing the `configure.ac` file; around line `781`, you'll find the checks for `libzip`.  If you comment these out and remove them, and then re-generate the `configure` script with the command `autoreconf -if`, you might be able to get past this.

Not *ideal*, I know, but it will be the quickest way.  I'll try to get a fix for this in for the upcoming `v4.9.3` release.  `v4.9.3` will be out as soon as possible, currently I am trying to figure out an issue in `nc_test` that has appeared with the latest OS update.  Thanks!

**Reply by vbsteph** [reply]
Thank you for looking into it! We will keep an eye on the next release, and I will try that workaround in the meantime.

**Reply by vbsteph** [reply]
It looks like the workaround you suggested will tide us over until the next release. Thank you for your help!

**Comment by WardF** [comment]
I've opened a PR here: https://github.com/Unidata/netcdf-c/pull/2793

**Comment by edwardhartnett** [comment]
We specifically require that zlib always be present for netCDF-4.

That is, we do not cope well without zlib. So there is no built-in handling of netcdf without zlib. This is really a mistake on my part, I just never understood (until quite recently) that HDF5 could be built without zlib. But it can.

So making it work might take a little more effort than you expect...

**Reply by vbsteph** [reply]
If I remember correctly, it was clear from the documentation that netCDF-4 requires zlib, but fortunately, our application doesn't require HDF5 support (yet). I had hoped that disabling it would eliminate the dependency requirement. 

If there is a way to support HDF5 without zlib dependence, I would be interested in seeing it. That said, I realize our use case isn't exactly common and that it might not be worth the time and effort to make something like that happen.

**Reply by edwardhartnett** [reply]
OK, if HDF5 is not being used, and nczarr is not being used, then I agree zlib should not be needed.

**Reply by WardF** [reply]
Just to clarify, this is about `libzip`, not `zlib`.

**Reply by WardF** [reply]
@vbsteph can you actually confirm my previous statement? I inferred from your original comments that we were talking about `libzip`, but I see now you've also mentioned `zlib`. I want to make sure I'm addressing the correct thing for the environment you're working in :)

**Reply by WardF** [reply]
Insofar as `zlib` is used to provide the `deflate` functionality found in hdf4/hdf5, and `libzip` was checked for unconditionally, but used for `ncZarr`-related functionality.

**Reply by vbsteph** [reply]
Yes, my question was about libzip, not zlib. I was under the impression that libzip uses zlib under the hood, so they are linked that way in my mind, but I'm certainly not an expert and could be wrong.

**Reply by WardF** [reply]
`libzip` may or may not use `zlib`, certainly; currently, we check to see if libzip is present before we check to see if we need it for use with `ncZarr`.  My PR changes that.  If we disable `ncZarr` but still have `hdf4` and `hdf5` functionality, `configure` will still check for and link against `zlib`.  Or at least, it should, and appears to in my testing.  Insofar as I can tell, the `PR` successfully only links against `libzip` if `ncZarr` is enabled.  I'll double check what is linked against when disabling things comprehensively, as you have done. But insofar as we wanted to eliminate an unnecessary linkage against `libzip`, I think we've got that :).

**Reply by vbsteph** [reply]
I don't _think_ zlib dependence would cause any issues with our setup, but we require neither nczarr nor hdf5 support, so my understanding is that both libzip and zlib would get eliminated from the dependency list. (Please correct me if I'm wrong.)

That said, it sounds like your changes should do exactly what we need, and I can always come back with separate questions about zlib if anything comes up later.

**Comment by DennisHeimbigner** [comment]
But as far as I know, HDF5 does not use libzip, it only requires zlib.
I think this issue is related to this PR https://github.com/Unidata/netcdf-c/pull/2712
and this PR https://github.com/Unidata/netcdf-c/pull/2722
